openedx/core/djangoapps/waffle_utils/__init__.py

```diff
@@ -1,503 +1,65 @@
-
-    WAFFLE_SWITCHES.is_enabled(waffle.ESTIMATE_FIRST_ATTEMPTED)
-
-To test WaffleSwitchNamespace, use the provided context managers.  For example:
-
-    with WAFFLE_SWITCHES.override(waffle.ESTIMATE_FIRST_ATTEMPTED, active=True):
-        ...
-
-For long-lived flags, you may want to change the default for devstack, sandboxes,
-or new Open edX releases. For help with this, see:
-openedx/core/djangoapps/waffle_utils/docs/decisions/0001-refactor-waffle-flag-default.rst
-
-Also see ``WAFFLE_FLAG_CUSTOM_ATTRIBUTES`` and docstring for _set_waffle_flag_attribute
-for temporarily instrumenting/monitoring waffle flag usage.
-
+Extra utilities for waffle: most classes are defined in edx_toggles.toggles (https://edx-toggles.readthedocs.io/), but
+we keep here some extra classes for usage within edx-platform. These classes cover course override use cases.
 """
-
 import logging
-import re
-import traceback
-from abc import ABCMeta
-from contextlib import contextmanager
-from weakref import WeakSet
 
-import crum
-import six
-from django.conf import settings
-from edx_django_utils.monitoring import set_custom_attribute
 from opaque_keys.edx.keys import CourseKey
-from waffle import flag_is_active, switch_is_active
 
-from openedx.core.lib.cache_utils import get_cache as get_request_cache
+from edx_toggles.toggles import WaffleFlag, WaffleFlagNamespace, WaffleSwitch, WaffleSwitchNamespace
 
 log = logging.getLogger(__name__)
 
 
-class WaffleNamespace(six.with_metaclass(ABCMeta, object)):
-    """
-    A base class for a request cached namespace for waffle flags/switches.
-
-    An instance of this class represents a single namespace
-    (e.g. "course_experience"), and can be used to work with a set of
-    flags or switches that will all share this namespace.
-    """
-
-    def __init__(self, name, log_prefix=None):
-        """
-        Initializes the waffle namespace instance.
-
-        Arguments:
-            name (String): Namespace string appended to start of all waffle
-                flags and switches (e.g. "grades")
-            log_prefix (String): Optional string to be appended to log messages
-                (e.g. "Grades: "). Defaults to ''.
-
-        """
-        assert name, "The name is required."
-        self.name = name
-        self.log_prefix = log_prefix if log_prefix else ''
-
-    def _namespaced_name(self, setting_name):
-        """
-        Returns the namespaced name of the waffle switch/flag.
-
-        For example, the namespaced name of a waffle switch/flag would be:
-            my_namespace.my_setting_name
-
-        Arguments:
-            setting_name (String): The name of the flag or switch.
-        """
-        return u'{}.{}'.format(self.name, setting_name)
-
-    @staticmethod
-    def _get_request_cache():
-        """
-        Returns a request cache shared by all instances of this class.
-        """
-        return get_request_cache('WaffleNamespace')
-
-
-class WaffleSwitchNamespace(WaffleNamespace):
```

Comment:
Well, deprecating is not a bad idea; it's just that I figured it was not strictly necessary in this case, as we are dealing with test utility functions. If you prefer I can add back an `.override()` method with a deprecation warning.

Reply:
It's the classes that I am more concerned with, and not the test method.  We could just `set_custom_attribute` in the init method, and follow-up with removal when we see it really isn't being used.

Note: I've been involved in several rollbacks in Production where we removed before deprecating, we thought we had searched everywhere, and things were missed in other repos we weren't aware of, sometimes outside of the edx org.

Reply:
I'm sorry, I'm not very familiar with the conventions of `set_custom_attribute`. What arguments should be passed to this function so that the warnings land in your dashboards? Something like `set_custom_attribute("deprecation", "WaffleFlag")`?

Reply:
I will manually search for these.  One possibility is:
```
set_custom_attribute('deprecated_waffle_flag', True)
set_custom_attribute('deprecated_waffle_switch', True)
```
I would need to do more searching this way, but if multiple things were used in the same call, I would see that.  Another option is:
```
set_custom_attribute('deprecated_waffle_utils', 'WaffleFlag')
set_custom_attribute('deprecated_waffle_utils', 'WaffleSwitch')
``
The second is simpler to query, so let's go with it.  The only downside is that the values would overwrite each other, in the event that multiple overridden classes were used during the same transaction/call.